### PR TITLE
 Extend multisig contract with weights

### LINKED ISSUE #285: Extend multisig contract with weights
We want to allow for election scenarios to happen off-chain. In many cases, different actors will have different voting power (just like validators).

We can do much of the vote aggregation off-chain, and use a multi-sig contract to control execution. However, it is missing the ability to provide different voting power to different keys.

Two proposal:
* Extend multisig to allow variable voting power per key (default to 1 if not defined for backwards compatibility)
* Define a new "voting" contract that contains different weights per key (and maybe other features?)

**Decision**

We will update the existing multisig to hold a weight per voter. However, there is no need for backwards compatibility if we can get this into a 0.12.x on zebranet, as so far we have no client code outside of this repo that uses the multi-sig feature.

- [x] Add a weight (uint8?) value along with each address, in all models and messages for multisig
- [x] Support reading this weight from genesis file as well
- [x] Enforce a range for weights, 0 < x < 256 (reject any tx/init with missing weights, weight=0)
- [x] Update all logic to use the weights to calculate if it meets threshold
- [ ] Note breaking change in the CHANGELOG

### Technical Spec
* Weight: 1-255 
* ActivationThreshold: > 0 && <= sum of all weighted signatures
* AdminThreshold: >0 < 2^63 ( can be > total sum of weighted signatures to lock contract)
* Max signatures in a contract: 100 to prevent cpu burning
* Cost is proportional to number of signatures
* Contract must have 1 member with weight > 0

To think about...

Currently the `multisig.UpdateContractMsg` behaves like a `POST` and overwrites the existing contract with a new state. All fields must be set like the original Create. However, the zero-value is illegal for all fields. We could modify this to `PATCH` and `Update` will only overwrite the fields defined. This means:

1. Admin and Activation Threshold of 0 means "don't change"
2. No Sigs means "don't change"

If there are Sigs, we must clarify what that means... overwrite current set with that new set? or do a diff, like with validators... setting weight=0 to mark current members for removal, and positive weight to update or add a member.

We could also allow `CreateContractMsg`, `UpdateContractMsg` (POST) and `PatchContractMsg` (ony apply diff.

(Read below in comments to see some example usages from @alpe 's doc)

### PR DESCRIPTION
**Pull request by husio**
`x/multisig` extension was updated and now each signature is having a
weight that defines what is the power of that signature.

- each signature power/weight must be greater than 0,
- activation threshold must be less or equal to the sum of all signature
  powers,
- admin threshold can be greater than the sum of all signature powers.
  This can be used to create a contract that can be only activated but
  never changed,
- multisig contract must have at least one participant/signature
  declared,
- maximum number of participants is 100 in order to prevent CPU burning,

- [ ] :bell: Update the CHANGELOG and include the breaking change description. 

resolve #285

### CODE REVIEW

File: x/multisig/codec.proto [hunk 1]
```diff
@@ -5,30 +5,35 @@ package multisig;
 import "github.com/gogo/protobuf/gogoproto/gogo.proto";
 
 message Contract {
-  // addresses to control it
-  repeated bytes sigs = 1;
-  // threshold needed to sign to activate it
-  int64 activation_threshold = 2;
-  // threshold needed to sign to change it
-  int64 admin_threshold = 3;
+  // Participants defines a list of all signatures that are allowed to sign the
```
Comment (ethanfrey) [inline x/multisig/codec.proto]: Nice protobuf definitions in any case

**Comment by ethanfrey** [comment]
Is this ready for a review?
I have plenty of other issues to go through, but ping me when you think you are ready (mention here or on slack) and I will give a proper pr review

**Comment by husio** [comment]
> Is this ready for a review?

Almost. I want to write a few more tests and double check that everything is implemented as expected. I will let you know when the code is ready for review.

File: x/multisig/decorator.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ func NewDecorator(auth x.Authenticator) Decorator {
 // Check enforce multisig contract before calling down the stack
 func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Checker) (weave.CheckResult, error) {
 	var res weave.CheckResult
-	newCtx, err := d.withMultisig(ctx, store, tx)
+	newCtx, err := d.authMultisig(ctx, store, tx)
```
Comment (ethanfrey) [inline x/multisig/decorator.go]: We could increment the cost by eg. 1 per participant in each contract referenced

File: x/multisig/decorator.go [hunk 3]
```diff
@@ -33,62 +33,49 @@ func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, ne
...
+		if err != nil {
+			return ctx, err
+		}
 
-			// check sigs (can be sig or multisig)
-			authenticated := x.HasNAddresses(ctx, d.auth, sigs, int(contract.ActivationThreshold))
-			if !authenticated {
-				return ctx, errors.Wrapf(errors.ErrUnauthorized, "contract=%X", contractID)
+		var power Weight
+		for _, p := range contract.Participants {
```
Comment (ethanfrey) [inline x/multisig/decorator.go]: Note this is expensive based on number of *possible* participant, not even signatures,
make sure there is a reasonable limit on number of participants, or at least a cost based on that. (see comment above)

File: x/multisig/decorator.go [hunk 4]
```diff
@@ -33,62 +33,49 @@ func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, ne
...
-	return ctx, nil
-}
-
-func (d Decorator) getContract(store weave.KVStore, id []byte) (*Contract, error) {
-	obj, err := d.bucket.Get(store, id)
-	if err != nil {
-		return nil, err
-	}
+		if power < contract.ActivationThreshold {
+			return ctx, errors.Wrapf(errors.ErrUnauthorized, "%d power is not enough", power)
```
Comment (ethanfrey) [inline x/multisig/decorator.go]: nice message. maybe include contract id as well?

File: x/multisig/decorator_test.go [hunk 5]
```diff
@@ -51,92 +61,71 @@ func TestDecorator(t *testing.T) {
...
-		err     error
+		wantErr *errors.Error
 	}{
-		// doesn't support multisig interface
-		{
-			&weavetest.Tx{Msg: &weavetest.Msg{Serialized: []byte{1, 2, 3}}},
-			[]weave.Condition{a},
-			nil,
-			nil,
+		"does not support multisig interface": {
```
Comment (ethanfrey) [inline x/multisig/decorator_test.go]: good cleanup here

File: x/multisig/decorator_test.go [hunk 6]
```diff
@@ -176,3 +165,18 @@ var _ weave.Tx = ContractTx{}
 func (p ContractTx) GetMultisig() [][]byte {
 	return p.MultisigID
 }
+
+func createContract(t *testing.T, db weave.KVStore, msg CreateContractMsg) []byte {
+	k := weavetest.NewCondition()
```
Comment (ethanfrey) [inline x/multisig/decorator_test.go]: Bit confused why we need k here.

File: x/multisig/model.go [hunk 7]
```diff
@@ -13,32 +13,42 @@ const (
...
 var _ orm.CloneableData = (*Contract)(nil)
 
-// Validate enforces sigs and threshold boundaries
 func (c *Contract) Validate() error {
-	if len(c.Sigs) == 0 {
-		return errors.Wrap(errors.ErrInvalidMsg, "missing sigs")
+	switch n := len(c.Participants); {
+	case n == 0:
+		return errors.Wrap(errors.ErrInvalidMsg, "no participants")
+	case n > 100:
```
Comment (ethanfrey) [inline x/multisig/model.go]: Nice check here.
Can you pull this into a contant (top of this file), like `const MaxParticipants`?

File: x/multisig/msg.go [hunk 8]
```diff
@@ -44,19 +36,62 @@ func (UpdateContractMsg) Path() string {
...
-	if c.ActivationThreshold <= 0 || int(c.ActivationThreshold) > len(c.Sigs) {
-		return errors.Wrap(errors.ErrInvalidMsg, invalidThreshold)
-	}
-	if c.AdminThreshold <= 0 {
-		return errors.Wrap(errors.ErrInvalidMsg, invalidThreshold)
+	return validateWeights(errors.ErrInvalidMsg,
+		c.Participants, c.ActivationThreshold, c.AdminThreshold)
+}
+
+// validateWeights returns an error if given participants and thresholds
```
Comment (ethanfrey) [inline x/multisig/msg.go]: Very nice and thorough set of checks.
I know the validate functions are tedious, but I think it is very important to be meticulous here, as they cut off many attack vectors as well as user errors.

File: x/multisig/msg.go [hunk 9]
```diff
@@ -20,21 +19,14 @@ func (CreateContractMsg) Path() string {
 
 // Validate enforces sigs and threshold boundaries
 func (c *CreateContractMsg) Validate() error {
-	if len(c.Sigs) == 0 {
-		return errors.Wrap(errors.ErrInvalidMsg, "missing sigs")
+	switch n := len(c.Participants); {
+	case n == 0:
+		return errors.Wrap(errors.ErrInvalidMsg, "no participants")
+	case n > 100:
```
Comment (ethanfrey) [inline x/multisig/msg.go]: Definitely use a constant, as this 100 is hardcoded a few places.

File: x/multisig/model.go [hunk 10]
```diff
@@ -13,32 +13,42 @@ const (
 	SequenceName = "id"
 )
 
-// enforce that Contract fulfils desired interface compile-time
+// Weight represents the strength of a signature.
+type Weight uint32
+
+func (w Weight) Validate() error {
```
Comment (ethanfrey) [inline x/multisig/model.go]: Not really sure why we need a type for one function.
Fine to keep it here, but I'd almost expect more logic with this.

File: x/multisig/msg.go [hunk 11]
```diff
@@ -44,19 +36,62 @@ func (UpdateContractMsg) Path() string {
...
-		return errors.Wrap(errors.ErrInvalidMsg, invalidThreshold)
+	return validateWeights(errors.ErrInvalidMsg,
+		c.Participants, c.ActivationThreshold, c.AdminThreshold)
+}
+
+// validateWeights returns an error if given participants and thresholds
+// configuration is not valid. This check is done on model and messages so
+// instead of copying the code it is extracted into this function.
+func validateWeights(
+	baseErr error,
```
Comment (ethanfrey) [inline x/multisig/msg.go]: Interesting that baseErr is only used for two of the cases (admin < activation and activation > total).
The others return same error from Validate.
Is this proper? (this is not a rhetorical question, I just want you to think about it and answer)

File: cmd/bnsd/scenarios/main_test.go [hunk 12]
```diff
@@ -159,7 +159,9 @@ func initGenesis(filename string, addr weave.Address) (*tm.GenesisDoc, error) {
 		},
 		"multisig": []interface{}{
 			dict{
-				"sigs":                 []weave.Address{addr},
+				"participants": []interface{}{
```
Comment (ethanfrey) [inline cmd/bnsd/scenarios/main_test.go]: :+1:

Comment (ethanfrey) [reply cmd/bnsd/scenarios/main_test.go]: But isn't there some other reference to multisig in cmd/bnsd?
I can't find it, but I thought there was something else Alex updated in AT tests.

File: x/multisig/decorator.go [hunk 13]
```diff
@@ -33,62 +33,49 @@ func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, ne
...
 
-			// load contract
-			contract, err := d.getContract(store, contractID)
-			if err != nil {
-				return ctx, err
-			}
+		// If already authenticated it does not matter if multisig can
+		// authenticate as well. Any authentication method is enough.
+		if d.auth.HasAddress(ctx, MultiSigCondition(contractID).Address()) {
+			return ctx, nil
```
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/decorator.go]: I think I did not understand this functionality correctly. 

I understand how multisig authorization with participant signatures works. Why would the multisig condition build from the contract ID be present in the context as part of the authentication?

File: x/multisig/decorator.go (hunk 2 above)
Comment (husio) [reply x/multisig/decorator.go]: Do you mean the gas cost? If yes then I can add it.

File: x/multisig/decorator.go (hunk 3 above)
Comment (husio) [reply x/multisig/decorator.go]: The limit is 100 as specified in the issue :man_scientist:

File: x/multisig/decorator.go (hunk 2 above)
Comment (ethanfrey) [reply x/multisig/decorator.go]: Yes, gas cost

File: x/multisig/decorator.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/multisig/decorator.go]: In the transaction, I can include a list of contracts I want to activate in order (eg. two sigs -> contract A -> contract B).

What we do when one of these addresses specified in the transaction is
a) nil
b) repeated (already in context).

I see two viable options:
1) skip this address as if the transaction didn't include it (auto-correct, suggested above)
2) return error as this case is invalid (harder, but will enforce client correctness)

stopping the processing process without an error (as if nothing else existed beyond this address) is not good. you can add a test with these two cases: `[contract1, contract1, contract2]` and `[contract1, nil, contract2]` and assert option 1 or 2 - your choice. Existing logic will activate `contract1` and return without error, which is incorrect in my mind.

File: x/multisig/decorator.go (hunk 3 above)
Comment (ethanfrey) [reply x/multisig/decorator.go]: note to self: rtfm

File: x/multisig/msg.go (hunk 11 above)
Comment (husio) [reply x/multisig/msg.go]: This is intended. If I am getting an error, I want to preserve it, so that it builds a nice chain of failures. Only if the condition is invalid and I have to create an error from scratch then I need some base error.
This function is used by both message and model validation so the base error differs.

I generally don't like those errors (`ErrInvalidMsg` and `ErrInvalidModel`) and I think we would be much better with a more generic `ErrInvalid` that covers all validation issues. Any detail will be part of the error message.

File: cmd/bnsd/scenarios/main_test.go (hunk 12 above)
Comment (husio) [reply cmd/bnsd/scenarios/main_test.go]: Those are acceptance tests :smile: 

This is the only one that was failing. I was not searching for more places where this could be used.

File: x/multisig/decorator_test.go (hunk 6 above)
Comment (husio) [reply x/multisig/decorator_test.go]: I do not fully understand this as this was here before :see_no_evil: 

If I remove the registration of a random condition for the purpose of authentication one test is failing:

```
--- FAIL: TestDecorator (0.00s)
    require.go:794: 
                Error Trace:    decorator_test.go:178
                                                        decorator_test.go:27
                Error:          Received unexpected error:
                            
                                github.com/iov-one/weave/x/multisig.CreateContractMsgHandler.validate
                                        /home/piotr/go/src/github.com/iov-one/weave/x/multisig/handlers.go:67
                                github.com/iov-one/weave/x/multisig.CreateContractMsgHandler.Deliver
                                        /home/piotr/go/src/github.com/iov-one/weave/x/multisig/handlers.go:43
                                github.com/iov-one/weave/x/multisig.createContract
                                        /home/piotr/go/src/github.com/iov-one/weave/x/multisig/decorator_test.go:177
                                github.com/iov-one/weave/x/multisig.TestDecorator
                                        /home/piotr/go/src/github.com/iov-one/weave/x/multisig/decorator_test.go:27
                                No signer: unauthorized
                Test:           TestDecorator
                Messages:       cannot deliver
FAIL
```

Without looking into details I think this might be because there is no signature authentication is failing without even trying. I will take time to understand this code :nerd_face: 
Is this an expected behaviour?

File: x/multisig/decorator.go [hunk 14]
```diff
@@ -33,62 +33,49 @@ func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, ne
...
+		return ctx, nil
+	}
 
-			// check if we already have it
-			if d.auth.HasAddress(ctx, MultiSigCondition(contractID).Address()) {
-				return ctx, nil
-			}
+	ids := multisigContract.GetMultisig()
+	for _, contractID := range ids {
+		if contractID == nil {
```
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/decorator.go]: Continued in https://github.com/iov-one/weave/pull/416#discussion_r267724373

File: x/multisig/decorator.go (hunk 2 above)
Comment (husio) [reply x/multisig/decorator.go]: Right now there is no gas cost modification in `x/multisig`. I will add one.

File: x/multisig/model.go (hunk 10 above)
Comment (alpe) [reply x/multisig/model.go]: I like it. A dedicated type is DDD style

File: x/multisig/msg.go [hunk 15]
```diff
@@ -44,19 +40,62 @@ func (UpdateContractMsg) Path() string {
...
 		}
 	}
+	if err := activationThreshold.Validate(); err != nil {
+		return errors.Wrap(err, "activation threshold")
+	}
+	if err := adminThreshold.Validate(); err != nil {
+		return errors.Wrap(err, "admin threshold")
+	}
+
+	var total Weight
```
Comment (alpe) [inline x/multisig/msg.go]: req: you need to check max power <256 as in the spec. Otherwise this can give you an overflow

File: x/multisig/decorator.go (hunk 2 above)
Comment (husio) [reply x/multisig/decorator.go]: @ethanfrey implemented in https://github.com/iov-one/weave/pull/416/commits/65a530b5ca991468699edb2f85368f2dcf7246b4

Please have a look if what I did there makes sense. I had to adjust the flow a bit but all tests are passing so I guess the functionality is still correctly implemented :crossed_fingers:

File: x/multisig/msg.go [hunk 16]
```diff
@@ -44,19 +40,62 @@ func (UpdateContractMsg) Path() string {
...
+	for _, p := range ps {
+		total += p.Power
```
Comment (alpe) [inline x/multisig/msg.go]: pp: optimization: you can exit loop early when activationThreshold is met

File: x/multisig/model.go (hunk 10 above)
Comment (husio) [reply x/multisig/model.go]: I did it for the [protobuf file to show that certain fields are about the same value](https://github.com/iov-one/weave/blob/65a530b5ca991468699edb2f85368f2dcf7246b4/x/multisig/codec.proto#L25). They are all of type `Weight` which makes you think about them as a single functionality.

I agree that using uint32 is less complex.

Do you think it is worth keeping the `Weight` type or this is an overkill for what it provides and using simple uint32 is enough?

File: x/multisig/decorator.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/decorator.go]: This was updated in https://github.com/iov-one/weave/commit/65a530b5ca991468699edb2f85368f2dcf7246b4

File: x/multisig/decorator.go [hunk 17]
```diff
@@ -33,62 +33,50 @@ func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, ne
...
+		}
 
-			// check sigs (can be sig or multisig)
-			authenticated := x.HasNAddresses(ctx, d.auth, sigs, int(contract.ActivationThreshold))
-			if !authenticated {
-				return ctx, errors.Wrapf(errors.ErrUnauthorized, "contract=%X", contractID)
+		var power Weight
+		for _, p := range contract.Participants {
+			if d.auth.HasAddress(ctx, p.Signature) {
+				power += p.Power
```
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/decorator.go]: True.
I need it for the gas as well, which is added now :)

File: x/multisig/decorator_test.go [hunk 18]
```diff
@@ -176,3 +165,18 @@ var _ weave.Tx = ContractTx{}
...
+	ctx := context.Background()
+	ctx = weave.WithHeight(ctx, 100)
+	auth := &weavetest.CtxAuth{Key: "authKey"}
+	ctx = auth.SetConditions(ctx, k)
+	handler := CreateContractMsgHandler{
+		auth:   auth,
+		bucket: NewContractBucket(),
+	}
+	res, err := handler.Deliver(ctx, db, &weavetest.Tx{Msg: &msg})
+	require.NoError(t, err, "cannot deliver")
```
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/decorator_test.go]: This is not my code :see_no_evil:

File: x/multisig/handlers.go [hunk 19]
```diff
@@ -125,44 +125,38 @@ func (h UpdateContractMsgHandler) Deliver(ctx weave.Context, db weave.KVStore, t
...
-	var sigs []weave.Address
-	for _, sig := range contract.Sigs {
-		sigs = append(sigs, sig)
-	}
-
-	// check sigs
-	authenticated := x.HasNAddresses(ctx, h.auth, sigs, int(contract.AdminThreshold))
-	if !authenticated {
-		return nil, errors.Wrapf(errors.ErrUnauthorized, "contract=%X", updateContractMsg.Id)
+	if err := msg.Validate(); err != nil {
```
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/handlers.go]: I moved it after checking the weights because I want to get `ErrUnauthorized` before any validation is executed. Otherwise, you can create a transaction without signing that will be validated.
Not sure if that is a bit deal but it feels more appropriate to check permissions before running any functionality.
What do you think?

File: x/multisig/init.go [hunk 20]
```diff
@@ -12,22 +12,28 @@ var _ weave.Initializer = (*Initializer)(nil)
...
+		ps := make([]*Participant, 0, len(c.Participants))
+		for _, p := range c.Participants {
+			ps = append(ps, &Participant{
+				Signature: p.Signature,
+				Power:     p.Power,
+			})
 		}
 		contract := Contract{
-			Sigs:                sigs,
+			Participants:        ps,
```
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/init.go]: Validation is called always by the bucket before persisting. It is not possible to save an entity that returns a validation error.

Does it make sense to additionally explicitly validate it?

File: x/multisig/msg.go (hunk 15 above)
Comment (husio) [reply x/multisig/msg.go]: Weight is uint32 as there is nothing below 32 bit in protobuf :muscle:

File: x/multisig/model.go (hunk 10 above)
Comment (ethanfrey) [reply x/multisig/model.go]: It's fine design.
We use custom type aliases for many different strings in iov-core for example.
But I hadn't seen such use in go/protobuf, mainly when needed to add functionality (like Address).

It's fine with me, just wanted to understand the reasoning

File: x/multisig/decorator_test.go [hunk 21]
```diff
@@ -176,3 +165,18 @@ var _ weave.Tx = ContractTx{}
...
+	k := weavetest.NewCondition()
+	ctx := context.Background()
+	ctx = weave.WithHeight(ctx, 100)
+	auth := &weavetest.CtxAuth{Key: "authKey"}
+	ctx = auth.SetConditions(ctx, k)
+	handler := CreateContractMsgHandler{
+		auth:   auth,
+		bucket: NewContractBucket(),
+	}
+	res, err := handler.Deliver(ctx, db, &weavetest.Tx{Msg: &msg})
```
[the remark this answers is not in the document]
Comment (husio) [reply x/multisig/decorator_test.go]: Yes, great spot! Done in https://github.com/iov-one/weave/pull/416/commits/fd1a9bf5f31b824b8edbbdaa5f1b8834cbe13c46

File: x/multisig/init.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/multisig/init.go]: No need. I agree this happens automatically in the Bucket.Save().
We should document this better (and weave in general)

File: x/multisig/decorator_test.go (hunk 6 above)
Comment (husio) [reply x/multisig/decorator_test.go]: After I have realized what it was doing, the code was updated. As @alpe pointed out in https://github.com/iov-one/weave/pull/416#discussion_r267768537 it can be done much better.

File: x/multisig/msg.go (hunk 15 above)
Comment (husio) [reply x/multisig/msg.go]: Done in https://github.com/iov-one/weave/pull/416/commits/21a8610af61bd88e5c958ce7bee05ea7f934dc0a

File: x/multisig/handlers.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (alpe) [reply x/multisig/handlers.go]: With other handlers like cash or escrow we do input validation before auth checks. I have not checked the others but it would be nice to be consistent

[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/multisig/handlers.go]: I agree. We should validate the message first.
But not modify db state until we have auth
